src/hooks/sync-sites-context.tsx

```diff
@@ -0,0 +1,212 @@
+					...prevStates,
+					[ remoteSiteId ]: {
+						...prevStates[ remoteSiteId ],
+						status: pullStatesProgressInfo.failed,
+					},
+				} ) );
+				getIpcApi().showErrorMessageBox( {
+					title: __( 'Failed to pull backup' ),
+					message: error instanceof Error ? error.message : __( 'Unknown error' ),
+				} );
+			}
+		},
+		[ __, client, pullStatesProgressInfo ]
+	);
+
+	const onBackupCompleted = useCallback(
+		async ( remoteSiteId: number, downloadUrl: string, selectedSite: SiteDetails ) => {
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					...prevStates[ remoteSiteId ],
+					status: pullStatesProgressInfo.completed,
+					downloadUrl,
+				},
+			} ) );
+
+			const filePath = await getIpcApi().downloadSyncBackup( remoteSiteId, downloadUrl );
+
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					...prevStates[ remoteSiteId ],
+					status: pullStatesProgressInfo.importing,
+					downloadUrl,
+				},
+			} ) );
+
+			await importFile(
+				{
+					path: filePath,
+					type: 'application/tar+gzip',
+				},
+				selectedSite
+			);
+
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					...prevStates[ remoteSiteId ],
+					status: pullStatesProgressInfo.finished,
+				},
+			} ) );
+			setTimeout( () => {
+				setPullStates( ( prevStates ) => {
+					const newStates = { ...prevStates };
+					delete newStates[ remoteSiteId ];
+					return newStates;
+				} );
+			}, 1000 );
+		},
+		[
+			importFile,
+			pullStatesProgressInfo.completed,
+			pullStatesProgressInfo.finished,
+			pullStatesProgressInfo.importing,
+		]
+	);
+
+	const getBackup = useCallback(
+		async ( remoteSiteId: number ) => {
+			if ( ! client ) {
+				return;
+			}
+			const backupId = pullStates[ remoteSiteId ]?.backupId;
+			if ( ! backupId ) {
+				console.error( 'No backup ID found' );
+				return;
+			}
+			const response = await client.req.get< {
+				status: 'in-progress' | 'completed' | 'failed';
+				download_url: string;
+			} >( `/sites/${ remoteSiteId }/studio-app/sync/backup`, {
+				apiNamespace: 'wpcom/v2',
+				backup_id: backupId,
+			} );
+
+			const statusWithProgress =
+				pullStatesProgressInfo[ response.status ] || pullStatesProgressInfo.failed;
+			const hasBackupCompleted = response.status === 'completed';
+			const downloadUrl = hasBackupCompleted ? response.download_url : null;
+
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					...prevStates[ remoteSiteId ],
+					status: statusWithProgress,
+					downloadUrl,
+				},
+			} ) );
+
+			if ( hasBackupCompleted && downloadUrl ) {
+				// Replacing the 'in-progress' status will stop the active listening for the backup completion
+				const selectedSite = pullStates[ remoteSiteId ]?.selectedSite;
+				if ( ! selectedSite ) {
+					return;
+				}
+				onBackupCompleted( remoteSiteId, downloadUrl, selectedSite );
+			}
```

Comment:
```suggestion
			if ( hasBackupCompleted && downloadUrl ) {
				// Replacing the 'in-progress' status will stop the active listening for the backup completion
				const selectedSite = pullStates[ remoteSiteId ]?.selectedSite;
				if ( ! selectedSite ) {
					return;
				}
				onBackupCompleted( remoteSiteId, downloadUrl, selectedSite );
			} else {
				setPullStates( ( prevStates ) => ( {
					...prevStates,
					[ remoteSiteId ]: {
						...prevStates[ remoteSiteId ],
						status: statusWithProgress,
						downloadUrl,
					},
				} ) );
			}
```
Nit, but this seems somewhat clearer to me, since `onBackupCompleted` will overwrite `pullStates` with new data immediately anyway.

Reply:
Thanks for the suggestion! https://github.com/Automattic/studio/pull/612/commits/e7387dfd984c6c0e755202addd863914814d9d4c